go/vt/vtctl/vtctl.go

```diff
@@ -2805,6 +2823,105 @@ func commandHelp(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.Flag
 	return nil
 }
 
+func commandVExec(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.FlagSet, args []string) error {
+	json := subFlags.Bool("json", false, "Output JSON instead of human-readable table")
+	dryRun := subFlags.Bool("dry_run", false, "Does a dry run of VExec and only reports the final query and list of masters on which it will be applied")
+	if err := subFlags.Parse(args); err != nil {
+		return err
+	}
+	if subFlags.NArg() != 2 {
+		return fmt.Errorf("usage: VExec --dry-run keyspace.workflow \"<query>\"")
+	}
+	keyspace, workflow, err := splitKeyspaceWorkflow(subFlags.Arg(0))
+	if err != nil {
+		return err
+	}
+	_, err = wr.TopoServer().GetKeyspace(ctx, keyspace)
+	if err != nil {
+		wr.Logger().Errorf("keyspace %s not found", keyspace)
+	}
+	query := subFlags.Arg(1)
+
+	results, err := wr.VExec(ctx, workflow, keyspace, query, *dryRun)
+	if err != nil {
+		return err
+	}
+	if *dryRun {
+		return nil
+	}
+	if len(results) == 0 {
+		wr.Logger().Printf("no result returned\n")
+	}
+	qr := queryResultFromVexecResults(results)
+	if *json {
+		return printJSON(wr.Logger(), qr)
+	}
+
+	printQueryResult(loggerWriter{wr.Logger()}, qr)
+	return nil
+}
+
+func queryResultFromVexecResults(results map[*topo.TabletInfo]*sqltypes.Result) *sqltypes.Result {
+	var qr *sqltypes.Result = &sqltypes.Result{}
+	qr.RowsAffected = uint64(len(results))
+	qr.Fields = []*querypb.Field{{
+		Name: "Tablet",
+		Type: sqltypes.VarBinary,
+	},
+		{
+			Name: "Rows Affected",
+			Type: sqltypes.Uint64,
+		},
+	}
+	for _, result := range results {
+		fields := result.Fields
+		qr.Fields = append(qr.Fields, fields...)
+		break
+	}
+	for tablet, result := range results {
+		var row2 []sqltypes.Value
+		row2 = append(row2, sqltypes.NewVarBinary(tablet.AliasString()))
+		row2 = append(row2, sqltypes.NewUint64(qr.RowsAffected))
+		for _, row := range result.Rows {
+			row2 = append(row2, row...)
```

Comment:
This may not work correctly if more than one row is returned. I see below that we support select. So, we should see what happens if a select returns more than one row.

Reply:
Fixed bug and improved reporting and tested more.